File: javascript/3212.js
var numberOfSubmatrices = function(grid) {
    const up_map = grid[0].map(_ => {return {X: 0, Y: 0, ".": 0}})
    return process_grid(grid, up_map, 0, 0) 
};

function process_grid(grid, up_map, idx, acc) {
    if (idx === grid.length) return acc
    const row_map = process_row(grid[idx], 0, {X: 0, Y: 0, ".": 0}, [])
    let c = 0
    new_up_map = []
    for (let i = 0; i < row_map.length; i++) {
        const one_map = {}
        one_map["X"] = row_map[i]['X'] + up_map[i]['X']
        one_map["Y"] = row_map[i]['Y'] + up_map[i]['Y']
        one_map["."] = row_map[i]['.'] + up_map[i]['.']
        new_up_map[i] = one_map
        if (one_map["X"] > 0 && one_map["Y"] === one_map["X"]) c = c + 1
    }
    return process_grid(grid, new_up_map, idx+1, acc+c)
}

function process_row(row, idx, cur, acc) {
    if (idx === row.length) return acc
    const h = row[idx]
    const c = cur[row[idx]]
    const new_map = {}
    new_map['X'] = cur['X']
    new_map['Y'] = cur['Y']
    new_map['.'] = cur['.']
    new_map[h] = c+1
    acc.push(new_map) // TODO
    return process_row(row, idx+1, new_map, acc)
}

File: elixir/3212.ex
defmodule Solution do
  @spec number_of_submatrices(grid :: [[char]]) :: integer
  def number_of_submatrices(grid) do
    up_map = Enum.map((hd grid), fn _a -> Map.new() end)
    process_grid(grid, up_map, 0)
  end

  # [[char]] -> [map] -> num
  def process_grid([], _up_map, acc), do: acc
  def process_grid([h|t], up_map, acc) do
    row_map = process_row(h, Map.new(), [])
    new_up_map =
      Enum.zip_with([row_map, up_map], fn [a,b] -> Map.merge(a, b, fn _k, v1, v2 -> v1 + v2 end) end)
    c = Enum.count(new_up_map, fn m -> ok(m) end)
    process_grid(t, new_up_map, acc+c)
  end

  def ok(m) do
     Map.get(m, ?X, 0) > 0 && Map.get(m, ?Y, 0) == Map.get(m, ?X, 0)
  end

  # [char] -> map -> [map]
  def process_row([], _cur, acc), do: acc |> Enum.reverse()
  def process_row([h|t], cur, acc) do
    c = Map.get(cur, h, 0)
    new_map = Map.put(cur, h, c+1)
    process_row(t, new_map, [new_map|acc])
  end
end
